plugins/http_plugin/include/eosio/http_plugin/beast_http_listener.hpp

```diff
@@ -0,0 +1,140 @@
+
+            beast_http_listener& operator=(const beast_http_listener&) = delete;
+            beast_http_listener& operator=(beast_http_listener&&) = delete;
+
+            beast_http_listener(std::shared_ptr<eosio::chain::named_thread_pool> thread_pool, 
+                                std::shared_ptr<ssl::context> ctx, 
+                                std::shared_ptr<http_plugin_state> plugin_state) : 
+                is_listening_(false)
+                , thread_pool_(std::move(thread_pool))
+                , ctx_(std::move(ctx))
+                , plugin_state_(std::move(plugin_state))
+                , acceptor_(thread_pool_->get_executor())
+                , socket_(thread_pool_->get_executor())
+            { }
+
+            virtual ~beast_http_listener() {
+                try {
+                     stop_listening();
+                } catch( ... ) { }
+            }; 
+
+            void listen(typename protocol_type::endpoint endpoint) {
+                if (is_listening_) return;
+
+                // for unix sockets we should delete the old socket
+                if(std::is_same<socket_type, stream_protocol::socket>::value) { 
+                    ::unlink(get_endpoint_path(endpoint).c_str());
+                }
+
+                beast::error_code ec;
+                // Open the acceptor
+                acceptor_.open(endpoint.protocol(), ec);
+                if(ec)  {
+                    fail(ec, "open", plugin_state_->logger, "closing port");
+                    return;
+                }
+
+                // Allow address reuse
+                acceptor_.set_option(asio::socket_base::reuse_address(true), ec);
+                if(ec)  {
+                    fail(ec, "set_option", plugin_state_->logger, "closing port");
+                    return;
+                }
+
+                // Bind to the server address
+                acceptor_.bind(endpoint, ec);
+                if(ec)  {
+                    fail(ec, "bind", plugin_state_->logger, "closing port");
+                    return;
+                }
+
+                // Start listening for connections
+                auto max_connections = asio::socket_base::max_listen_connections;
+                fc_ilog( plugin_state_->logger, "acceptor_.listen()" ); 
+                acceptor_.listen(max_connections, ec);
+                if(ec) {
+                    fail(ec, "listen", plugin_state_->logger, "closing port");
+                    return;
+                }
+                is_listening_ = true;        
+            }
+
+            // Start accepting incoming connections
+            void start_accept()
+            {
+                if(!is_listening_) return;
+                do_accept();
+            }
+
+            bool is_listening() {
+                return is_listening_;
+            }
+
+            void stop_listening() {
+                if(is_listening_) {
+                    thread_pool_->stop();
```

Comment:
http_plugin::plugin_shutdown() is calling stop() on the thread_pool already. Doesn't seem appropriate to do it here, as http_plugin/impl is seemingly the "owner" of the thread pool since it created it